x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/RandomizedAssignmentRounding.java

```diff
@@ -0,0 +1,428 @@
+ * double values for the number of allocations each model gets on each node, as well as a double
+ * value in [0-1] for whether there is an assignment for each model to each node.
+ */
+class RandomizedAssignmentRounding {
+
+    private static final Logger logger = LogManager.getLogger(RandomizedAssignmentRounding.class);
+
+    private static final double EPS = 1e-6;
+
+    private final Random random;
+    private final int rounds;
+    private final Collection<Node> nodes;
+    private final Collection<Model> models;
+    private final AssignmentHolder assignmentHolder;
+
+    RandomizedAssignmentRounding(Random random, int rounds, Collection<Node> nodes, Collection<Model> models) {
+        if (rounds <= 0) {
+            throw new IllegalArgumentException("rounds must be > 0");
+        }
+        this.random = Objects.requireNonNull(random);
+        this.rounds = rounds;
+        this.nodes = Objects.requireNonNull(nodes);
+        this.models = Objects.requireNonNull(models);
+        this.assignmentHolder = new AssignmentHolder();
+    }
+
+    AssignmentPlan computePlan(Map<Tuple<Model, Node>, Double> allocationVars, Map<Tuple<Model, Node>, Double> assignmentVars) {
+        AssignmentPlan bestPlan = assignmentHolder.toPlan();
+
+        assignmentHolder.initializeAssignments(allocationVars, assignmentVars);
+        assignmentHolder.assignUnderSubscribedNodes();
+        List<Tuple<Model, Node>> softAssignmentQueue = assignmentHolder.createSoftAssignmentQueue();
+
+        if (softAssignmentQueue.isEmpty() == false) {
+            logger.debug(() -> new ParameterizedMessage("Random assignment rounding across [{}] rounds", rounds));
+            for (int i = 0; i < rounds; i++) {
+                AssignmentHolder randomizedAssignments = new AssignmentHolder(assignmentHolder);
+                randomizedAssignments.doRandomizedRounding(softAssignmentQueue);
+                AssignmentPlan randomizedPlan = randomizedAssignments.toPlan();
+                if (randomizedPlan.compareTo(bestPlan) > 0) {
+                    bestPlan = randomizedPlan;
+                }
+            }
+        } else {
+            AssignmentPlan plan = assignmentHolder.toPlan();
+            if (plan.compareTo(bestPlan) > 0) {
+                bestPlan = plan;
+            }
+        }
+
+        return bestPlan;
+    }
+
+    private class AssignmentHolder {
+        private final Map<Tuple<Model, Node>, Double> softAssignments = new HashMap<>();
+        private final Map<Tuple<Model, Node>, Double> softAllocations = new HashMap<>();
+        private final ResourceTracker resourceTracker;
+
+        private AssignmentHolder() {
+            resourceTracker = new ResourceTracker(nodes, models);
+        }
+
+        private AssignmentHolder(AssignmentHolder holder) {
+            softAssignments.putAll(holder.softAssignments);
+            softAllocations.putAll(holder.softAllocations);
```

Comment:
I find the naming of these a bit confusing since you convert them from soft into hard. One this basis I would prefer `assignments` and `allocations`.